bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+	 *            {@link DataSetMonitorWorker}
+	 */
+	public DataSetMonitorWorker(String bucketName, DataSetProcessor dataSetProcessor) {
+		this.bucketName = bucketName;
+		this.dataSetProcessor = dataSetProcessor;
+	}
+
+	/**
+	 * @see java.lang.Runnable#run()
+	 */
+	@Override
+	public void run() {
+		/*
+		 * Request a list of all objects in the configured bucket and directory.
+		 * (In the results, we'll be looking for the oldest manifest file, if
+		 * any.)
+		 */
+		ListObjectsRequest bucketListRequest = new ListObjectsRequest();
+		bucketListRequest.setBucketName(bucketName);
+
+		/*
+		 * S3 will return results in separate pages. Loop through all of the
+		 * pages, looking for the oldest manifest file that is available.
+		 */
+		String manifestToProcessKey = null;
+		ObjectListing objectListing = s3Client.listObjects(bucketListRequest);
+		do {
+			for (S3ObjectSummary objectSummary : objectListing.getObjectSummaries()) {
+				String key = objectSummary.getKey();
+				if (REGEX_MANIFEST.matcher(key).matches()) {
+					/*
+					 * We've got an object that *looks like* it might be a
+					 * manifest file. But we also need to ensure that it starts
+					 * with a valid timestamp, and if so, check to see if that
+					 * timestamp is the oldest one we've encountered so far. If
+					 * so, we mark it as the "current oldest" and continue
+					 * looking through the other keys.
+					 */
+					Instant dataSetTimestamp = parseDataSetTimestamp(key);
+					if (dataSetTimestamp != null && manifestToProcessKey == null)
+						manifestToProcessKey = key;
+					else if (dataSetTimestamp != null && manifestToProcessKey != null
+							&& (dataSetTimestamp.compareTo(parseDataSetTimestamp(manifestToProcessKey)) < 0))
+						manifestToProcessKey = key;
+				}
+			}
+
+			objectListing = s3Client.listNextBatchOfObjects(objectListing);
+		} while (objectListing.isTruncated());
+
+		// If no manifest was found, we're done (until next time).
+		if (manifestToProcessKey == null)
+			return;
+
+		// We've found the oldest manifest. Now go download and parse it.
+		DataSetManifest dataSetManifest = readManifest(manifestToProcessKey);
+
+		/*
+		 * We've got a dataset to process. However, it might still be uploading
+		 * to S3, so we need to wait for that to complete before we start
+		 * processing it.
+		 */
+		boolean dataSetComplete = false;
+		do {
+			if (dataSetIsAvailable(dataSetManifest))
+				dataSetComplete = true;
+
+			/*
+			 * We'll loop forever, but it's prudent to pause between each
+			 * iteration.
+			 */
+			try {
+				Thread.sleep(1000 * 1);
+			} catch (InterruptedException e) {
+				/*
+				 * Many Java applications use InterruptedExceptions to signal
+				 * that a thread should stop what it's doing ASAP. This app
+				 * doesn't, so this is unexpected, and accordingly, we don't
+				 * know what to do. Safest bet is to blow up.
+				 */
+				throw new RuntimeException(e);
+			}
+		} while (!dataSetComplete);
```

Comment:
Would it make sense to put some sort of threshold on how long it will wait before at least sending out some kind of notification that something may have gone wrong?